azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java

```diff
@@ -1774,144 +1784,177 @@ private void ajaxHandleUpload(final HttpServletRequest req, final HttpServletRes
-    } else if (project == null) {
-      registerError(ret, "Installation Failed. Project '" + projectName
-          + "' doesn't exist.", resp, 400);
-    } else if (!hasPermission(project, user, Type.WRITE)) {
-      registerError(ret, "Installation Failed. User '" + user.getUserId()
-          + "' does not have write access.", resp, 400);
-    } else {
-      ret.put("projectId", String.valueOf(project.getId()));
-
-      final FileItem item = (FileItem) multipart.get("file");
-      final String name = item.getName();
-      String type = null;
-
-      final String contentType = item.getContentType();
-      if (contentType != null
-          && (contentType.startsWith(APPLICATION_ZIP_MIME_TYPE)
-          || contentType.startsWith("application/x-zip-compressed") || contentType
-          .startsWith("application/octet-stream"))) {
-        type = "zip";
-      } else {
-        item.delete();
-        registerError(ret, "File type " + contentType + " unrecognized.", resp, 400);
-
-        return;
-      }
-
-      final File tempDir = Utils.createTempDir();
-      OutputStream out = null;
-      try {
-        logger.info("Uploading file " + name);
-        final File archiveFile = new File(tempDir, name);
-        out = new BufferedOutputStream(new FileOutputStream(archiveFile));
-        IOUtils.copy(item.getInputStream(), out);
-        out.close();
-
-        if (this.enableQuartz) {
-          //todo chengren311: should maintain atomicity,
-          // e.g, if uploadProject fails, associated schedule shouldn't be added.
-          this.scheduler.unschedule(project);
-        }
-
-        // get the locked flows for the project, so that they can be locked again after upload
-        List<String> lockedFlows = getLockedFlows(project);
+      return null;
+    }
+    if (!hasPermission(project, user, Type.WRITE)) {
+      registerError(ret,
+          "Installation Failed. User '" + user.getUserId() + "' does not have write access.",
+          resp, 400);
+      return null;
+    }
+    return project;
+  }
 
-        final Map<String, ValidationReport> reports =
-            this.projectManager.uploadProject(project, archiveFile, type, user,
-                props);
+  /**
+   * Remove schedule of renamed/deleted flows
+   *
+   * @param project project from which old flows will be unscheduled
+   * @param scheduleManager the schedule manager
+   * @param onDeletedSchedule a callback function to execute with every deleted schedule
+   */
+  static void removeScheduleOfDeletedFlows(final Project project,
+      final ScheduleManager scheduleManager, final Consumer<Schedule> onDeletedSchedule)
+      throws ScheduleManagerException {
+    final Set<String> flowNameList = project.getFlows().stream().map(f -> f.getId()).collect(
+        Collectors.toSet());
 
-        if (this.enableQuartz) {
-          this.scheduler.schedule(project, user.getUserId());
-        }
+    for (final Schedule schedule : scheduleManager.getSchedules()) {
```

Comment:
Would it be possible to get schedules by project, to limit the number of flows fetched?

Reply:
It looks like there is no direct relationship between Project and Schedule. The list of schedules is coming from memory and Im filtering out by project first to reduce work. Creating a getSchedulesByProject() method in ScheduleManager class will do essentially the same thing I'm doing.